### PR TITLE
Reload environment variables by default; add setting to disable

### LINKED ISSUE #1125: Feature Request: Terminal should listen for the WM_SETTINGCHANGE for environment variable updates
# Summary of the new feature/enhancement

When I install an app that updates the path or I update the path manually, it isn't enough to kill an individual tab in the Terminal and start a new one.  The new tab still inherits the old, unchanged environment variable values. This means I have to kill/restart the Terminal and lose all my tabs.

# Proposed technical implementation details (optional)

Add a Windows message handler for `WM_SETTINGCHANGE` and update the Terminal process's environment block so that any new tabs get the updated environment variables.

### PR DESCRIPTION
**Pull request by ianjoneill**
## Summary of the Pull Request
Adds a global setting `compatibility.reloadEnvironmentVariables` with a default value of `true`. When set, during connection creation a new environment block will be generated to ensure it has the latest environment variables.

## Validation Steps Performed
Manually tested - see video in the ~~first~~ second comment.

## PR Checklist
- [x] Closes #1125
- [ ] Tests added/passed
- [ ] Documentation updated
   - If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/terminal) and link it here: #xxx
- [x] Schema updated (if necessary)

### CODE REVIEW

File: src/cascadia/TerminalSettingsModel/MTSMSettings.h [hunk 1]
```diff
@@ -27,6 +27,7 @@ Author(s):
     X(bool, ForceFullRepaintRendering, "experimental.rendering.forceFullRepaint", false)                                                                   \
     X(bool, SoftwareRendering, "experimental.rendering.software", false)                                                                                   \
     X(bool, UseBackgroundImageForWindow, "experimental.useBackgroundImageForWindow", false)                                                                \
+    X(bool, ReloadEnvironmentVariables, "experimental.reloadEnvironmentVariables", true)                                                                   \
```
Comment (zadjii-msft) [inline src/cascadia/TerminalSettingsModel/MTSMSettings.h:30]:
```suggestion
    X(bool, ReloadEnvironmentVariables, "compatibility.reloadEnvironmentVariables", true)                                                                   \
```

File: doc/cascadia/profiles.schema.json [hunk 2]
```diff
@@ -2164,6 +2164,11 @@
           "description": "When set to true, the background image for the currently focused profile is expanded to encompass the entire window, beneath other panes.",
           "type": "boolean"
         },
+        "experimental.reloadEnvironmentVariables": {
```
Comment (zadjii-msft) [inline doc/cascadia/profiles.schema.json:2167]:
```suggestion
        "compatibility.reloadEnvironmentVariables": {
```

File: src/cascadia/TerminalControl/IControlSettings.idl [hunk 3]
```diff
@@ -61,5 +61,6 @@ namespace Microsoft.Terminal.Control
         Boolean SoftwareRendering { get; };
         Boolean ShowMarks { get; };
         Boolean UseBackgroundImageForWindow { get; };
+        Boolean ReloadEnvironmentVariables { get; };
```
Comment (zadjii-msft) [inline src/cascadia/TerminalControl/IControlSettings.idl:64]: honestly I think we can make the diff even smaller. I don't _think_ the control actually needs to know if `ReloadEnvironmentVariables` is set at all. You should be able to revert the changes to `IControlSettings`, `TerminalSettings.*`, and just check `_settings.GlobalSettings().ReloadEnvironmentVariables()` in `TerminalPage`

**Review by zadjii-msft — COMMENTED** [review]
wow you're right that was easy. 

My tiny nit is that we've been discussing internally on re-aligning some of the `experimental` settings we have as `compatibility` flags instead. That's not something you'd have possibly known, and I'm pretty sure my initial writeup called for `experimental.` too. 

Honestly this is gonna be such goodness that I want to just hit ✅ anyways

File: src/cascadia/TerminalControl/IControlSettings.idl (hunk 3 above)
Comment (ianjoneill) [reply src/cascadia/TerminalControl/IControlSettings.idl]: You're right. I saw a `settings` available to me in `TerminalPage` and thought "I'll use that" :)

**Comment by ianjoneill** [comment]
Umm why am I getting static analysis failures on code I haven't touched?

https://dev.azure.com/ms/terminal/_build/results?buildId=439511&view=results

**Comment by ianjoneill** [comment]
Now the build's failing due to #14581.

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 4]
```diff
@@ -1280,6 +1281,16 @@ namespace winrt::TerminalApp::implementation
...
+            {
+                til::env environment;
+                environment.regenerate();
+
+                for (const auto& [key, value] : environment.as_map())
+                {
+                    envMap.Insert(key, value);
+                }
+            }
             envMap.Insert(L"WT_PROFILE_ID", guidWString);
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TerminalPage.cpp:1294]: I love the simplicity of this, but I am a little bit concerned about what happens after we hand this environment block off.

We produce a new env block here, painstakingly inserting it into envMap... and then ConptyConnection uses `EnvironmentVariableMapW` to _generate a new map based on the current process environment_ and inserts the contents of `envMap` from here into it.

In brief... what we will get is a hybrid of Terminal's spawning environment and the current environment. As well, we will probably end up parsing the entire env twice.

Should we move this down to ConptyConnection (worm the boolean down in through the valueset for all I mind, ha) and make sure we do not call `UpdateEnvironmentMapW`?

**Review by DHowett — CHANGES_REQUESTED** [review]

**Comment by zadjii-msft** [comment]
Huh, moving that down to connection made audit run on `env.h`, which it does _not_ like. If you don't want to fix them now, I'm okay with that, there's a way to disable audit mode from yelling about an include. You could add that around the include and then we could just fix that in post. 

Probably like this, with a push/pop
```c++

// Some of the interactivity classes pulled in by ServiceLocator.hpp are not
// yet audit-safe, so for now we'll need to disable a bunch of warnings.
#pragma warning(disable : 26432)
#pragma warning(disable : 26440)
#pragma warning(disable : 26455)

// We end up including ApiMessage.h somehow, which uses nameless unions
#pragma warning(disable : 4201)

```

Or you could fix the warnings now :P
```
    35>C:\a\_work\1\s\src\inc\til\env.h(346): error C26490: Don't use reinterpret_cast (type.1). 
    35>C:\a\_work\1\s\src\inc\til\env.h(352): error C26490: Don't use reinterpret_cast (type.1). 
    35>C:\a\_work\1\s\src\inc\til\env.h(489): error C26440: Function 'til::env::is_path_var' can be declared 'noexcept' (f.6). 
    35>C:\a\_work\1\s\src\inc\til\env.h(521): error C26481: Don't use pointer arithmetic. Use span instead (bounds.1). 
    35>C:\a\_work\1\s\src\inc\til\env.h(506): error C26481: Don't use pointer arithmetic. Use span instead (bounds.1). 
    35>C:\a\_work\1\s\src\inc\til\env.h(526): error C26455: Default constructor should not throw. Declare it 'noexcept' (f.6). 
    35>C:\a\_work\1\s\src\inc\til\env.h(574): error C26440: Function 'til::env::as_map' can be declared 'noexcept' (f.6). 
    35>C:\a\_work\1\s\src\cascadia\TerminalConnection\ConptyConnection.cpp(106): error C26478: Don't use std::move on constant variables. (es.56). 
```

**Review by DHowett — APPROVED** [review]
beautiful. thanks for the quick patch!

**Comment by DHowett** [comment]
Hey @lhecker, would you be willing to fix the audit failures on-branch so we can quick merge Ian's work?

**Comment by lhecker** [comment]
I've fixed the audit mode issues, as well as some others. I've also made some further improvements to `til::env` along the way.

However now that I understand how `til::env` was intended to be used, I have to confess that I'd be happier with a solution like ConEmu's. Theirs is clearly superior as it's simpler and more performant. It only reloads on `WM_SETTINGCHANGE`, just like the feature request asked. Because of this I'm not sure if the exact feature request of #1125 is solved with this PR just yet. Alternatively we should create a followup issue to address this, as the code that `til::env` runs is rather heavy and there's almost never any reason to run it. (After all, how often do environment variables change anyways?)

**Comment by zadjii-msft** [comment]
In the name of incremental progress - merge now, and fast-follow with a PR for "cache a static 'base' env block and only reload it on WM_SETTINGCHANGE"?

**Comment by ianjoneill** [comment]
I had actually fixed some of the audit issues this morning - rather than just ignoring them :)

**Review by DHowett — APPROVED** [review]

**Comment by DHowett** [comment]
Ah, sorry to override/duplicate your effort @ianjoneill!

**Comment by Ares9323** [comment]
> ReloadEnvironmentVariables.mp4

I've tried to look for `"experimental.reloadEnviromentVariables": true,` or similar names both in the regular windows terminal and in the preview version but I wasn't able to find anything, has it been implemented differently or elsewhere?

**Comment by zadjii-msft** [comment]
What are you trying to accomplish exactly? The setting added as `compatibility.reloadEnvironmentVariables`, in the globals, but it's also enabled by default in 1.18.

**Comment by Ares9323** [comment]
> What are you trying to accomplish exactly? The setting added as `compatibility.reloadEnvironmentVariables`, in the globals, but it's also enabled by default in 1.18.

Thank you for your answer, I tried looking for both `experimental.reloadEnviromentVariables` and  `compatibility.reloadEnvironmentVariables` in the settings.json but nothing was showing up in the autocompletion, so I wrongly assumed that the option was missing (Screenshot from WT preview 1.18.14): 
image

The default terminal was obviously not working because it hasn't been updated to 1.18 yet:
image
